### ISSUE TITLE
Stack unwinding doest not run ZIO.Folds on recoverable errors after interruption

### ISSUE THREAD

**Issue by mvv** [scalaz/scalaz-zio#729]
```scala
for {
  r <- Ref.make("NO")
  f <- IO.never.ensuring {
      (IO.unit *> IO.fail("ERR")).catchAll(r.set)
    } .fork
  x <- f.interrupt
  v <- r.get
} yield (x, v)
```
returns `(Failure(Fail("ERR")), "NO")`. If I replace `(IO.unit *> IO.fail("ERR"))` with just `IO.fail("ERR")` the result becomes `(Failure(Interrupt), "YES")`.

There are two `unwindStack`s that happen after `f` is interrupted. First is triggered by the interruption itself and returns finalizer `(IO.unit *> IO.fail("ERR")).catchAll(r.set)`, the second happens when the execution reaches `IO.fail("ERR")`. At that point `allowRecovery` is `false` (due to `state.interrupted` and `state.terminating` both being `true`), which prevents `ZIO.Fold`s on stack from firing.

Basically, `unwindStack` and `allowRecovery` misinterpret the cause of stack unwinding to be interruption, ignoring the real cause.

**Comment by jdegoes** [comment]
Fixed.
